### DISCUSSION TITLE
v7.x compatibility question

### DISCUSSION THREAD

**Discussion by RwwL** [i18next/next-i18next#2181] [Q&A]
I'm working on an older app that's still on Next.js v10.x and `next-i18next` v7.x. 

I'm in the middle of migrating to Next.js v12.x now, and trying to get a sense of whether I absolutely have to upgrade to a more current `i18next` now or whether I should be able to keep things running on 7.x for awhile, possibly with some light refactoring, and migrate that part later, since there's already plenty of other things that need attention right away as part of the Next.js upgrade.

Right now the big sticking point with just upgrading `next-i18next` is the apparent need to add the `serverSideTranslations` call to every page since it's a large app (100+ pages). I'm still wrapping my head around our entire 7.x setup, but the current state of things in my local app now on 12.x is that direct-loads of pages seem to get all the required translation files through our old setup, but there are also new 404s for some of the locale `.json` files in the console. Client-rendered subsequent page loads in development seem to be missing all their translations. 

Any getting-started guidance I can get from the community on this is very welcome, thanks.

**Comment by adrai** [comment]
fyi: v7.x was before the major rewrite, that was leveraging the built-in internationalised routing provided by NextJs.

**Reply by RwwL** [reply]
Thanks, @adrai, I did remember reading that [here](https://github.com/i18next/next-i18next/releases/tag/v8.0.0) but I'm not sure that answers the question of whether 7.x can still work for a while with Next 12.x or if it's critical to upgrade at this particular point.

**Reply by adrai** [reply]
Sorry, never tried.

**Reply by RwwL** [reply]
@adrai I went ahead with the upgrade to v8, a little scripting took care of the `serverSideTranslations` adds and the basics are working great.

We do have one issue I can't figure out yet and I figured I'd tag you here, sorry to bug you again:

- there are two certain page types in our app that the user can custom-name
- the first time you go to any into the section that contains those two page types, we call an APi endpoint to load those custom names and call `i18n.addResources('en', 'hot-loaded', { ...newKeys }`, adding the new keys to a `hot-loaded` namespace
- every page in the section sets up this handler when the page mounts: 
  ```
  i18n.store.on('added', () => {
     i18n.changeLanguage('en');
  });
  ```
 - this has worked historically in our app, but now it works only on the first page view in that section, when you navigate to another page in the section it seems that the `hot-loaded` namespace has been lost.
 
Any ideas there?

**Reply by adrai** [reply]
with the new version you know need to add the hot-loaded resources in your [getStaticProps](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation) or [getServerSideProps](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering) function.
Something like this:

```js
export const getServerSideProps = async ({
  locale,
}) => {
  const p = await serverSideTranslations(locale ?? 'en', [
    'common',
    'footer',
  ])
  const newKeys = { hi: 'there' } // coming from somewhere...
  p._nextI18Next.initialI18nStore.en['hot-loaded'] = { ...newKeys }
  return {
    props: {
      ...p,
    },
  }
}
```

or alternatively do it on client side, like this: https://github.com/i18next/i18next-http-backend/blob/master/example/next/pages/lazy-reload-page.js#L16

**Reply by RwwL** [reply]
Yeah, I'll definitely need to be adding the hot-loaded resources on the client side in our use case, thanks. That example you linked seems to be more about consuming those hot-loaded resources though. I'm concerned that I'm not adding them correctly for v8.x+ anymore. After I've made a call to our API to get these custom section names, I'm adding them like this:

```
import { i18n } from 'next-i18next';

const getSpecialSectionTranslations = async () => {
   const data = await fetchFromAPI();
   i18n.addResources(
      'en',
      'hot-loaded',
      {
        [custom keys & values from API-fetched data]
      }
    );
} 
```

...but this isn't working anymore, apparently — at least, the setup at the point of consumption using `bindI18n` and `reloadResources` doesn't respond to the update when I call `addResources` above. I thought the problem was that I wasn't calling `addResources` on a proper `i18n` instance, so I tried passing it from the nearest component that's calling `const { t, i18n } = useTranslation()` but that didn't get it working either.

**Reply by adrai** [reply]
you need to bind the added event for that => `bindI18n: 'languageChanged added'`

**Reply by RwwL** [reply]
Thanks, but I'm already doing that at the point of consumption, per [your linked suggestion](https://github.com/i18next/i18next-http-backend/blob/master/example/next/pages/lazy-reload-page.js#L16). though. Does it need to happen in another place too (like, somewhere in the same file where I call `addResources`)?

Here's the hook we use for consuming the custom section names, (just anonymized a bit for company policy), this is what I'm calling the point of consumption:

```js
import { useTranslation } from 'next-i18next';
import { useEffect } from 'react';

const useCustomSectionTranslation = () => {
  const { t, i18n } = useTranslation(['hot-loaded'], { bindI18n: 'languageChanged loaded' });
  const term1Singular = t('customSection.term1Singular');
  const term1Plural = t('customSection.term1Plural');
  const term2Singular = t('customSection.term2Singular');
  const term2Plural = t('customSection.term2Plural');

  useEffect(() => {
   // the example used i18n.resolvedLanguage but that seems to always be undefined for me, language is not
    i18n.reloadResources(i18n.language, ['hot-loaded']);
  }, []); // I've also tried passing in i18n as a dependency to this useEffect
  return {
    term1Singular,
    term1Plural,
    term2Singular,
    term2Plural
  };
};

export default useCustomSectionTranslation;
```

**Reply by adrai** [reply]
Replace `languageChanged loaded` with `languageChanged added`

**Reply by RwwL** [reply]
So sorry to keep bugging you. I've replaced that too but I'm still stuck in the same situation. I'm able to get the custom keys rendered by that hook on the inital load of any given page in the section, but on subsequent navigations I lose them.

Something that's puzzling to me: this situation seems to be the same whether or not I supply the `hot-loaded` namespace in the page's call to `serverSideTranslation`. If it matters, we do this via `getStaticProps` and not `getServerSideProps` but I don't think it should.

**Reply by adrai** [reply]
not easy... would be nice to have a minimal reproducible example repo to investigate

**Reply by RwwL** [reply]
Yeah, I'll see if I can make the time to try to build that later. Before I do, can I ask one more question? It seems to me like this leftover piece of our old implementation might be obsolete and need to come out, would you agree?

```js
// Ensure that anytime a translation is programmatically updated
// all instances of i18next will re-render with the new data.
i18n.store.on('added', () => {
  i18n.changeLanguage('en');
});
```

Calling `changeLanguage` when we're not really changing languages feels a bit like a hack; did our previous developer set this up strangely? It seems that if I'm calling `addResources` in one place and then using `bindI18n: 'languageChanged added'` elsewhere then maybe that's all that should be needed?

But apparently not, because if I comment that out then the initial-page-load scenario stops working.

**Reply by adrai** [reply]
yes, probably a leftover piece

**Reply by RwwL** [reply]
Well, if I remove it, the only part that's currently working stops working (having the custom keys render on an initial page load in the section that uses them).

**Reply by adrai** [reply]
like said... a reproducible example would help

**Reply by RwwL** [reply]
Yeah, I know, thanks. But the time required to build that is probably more than the time it'd take to refactor our app to not use this `hot-loaded` namespace at all, and simply render the custom values when they exist, and render those strings via `useTranslation` when they don't. I'll probably go that route, but thanks again very much for your time (both on this thread and the general maintenance of these projects).
